pkg/compactor/compactor.go

```diff
@@ -0,0 +1,225 @@
+package compactor
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/gardener/etcd-backup-restore/pkg/compressor"
+	"github.com/gardener/etcd-backup-restore/pkg/metrics"
+	"github.com/gardener/etcd-backup-restore/pkg/miscellaneous"
+	"github.com/gardener/etcd-backup-restore/pkg/snapshot/restorer"
+	"github.com/gardener/etcd-backup-restore/pkg/snapstore"
+	brtypes "github.com/gardener/etcd-backup-restore/pkg/types"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	"go.etcd.io/etcd/clientv3"
+)
+
+const (
+	tmpDir                          = "/tmp"
+	defaultName                     = "default"
+	defaultInitialAdvertisePeerURLs = "http://localhost:2384"
+	defaultInitialClusterToken      = "etcd-cluster"
+	defaultMaxFetchers              = 6
+	defaultMaxCallSendMsgSize       = 10 * 1024 * 1024 //10Mib
+	defaultMaxRequestBytes          = 10 * 1024 * 1024 //10Mib
+	defaultMaxTxnOps                = 10 * 1024
+	defaultEmbeddedEtcdQuotaBytes   = 8 * 1024 * 1024 * 1024 //8Gib
+	etcdDialTimeout                 = time.Second * 30
+	etcdDir                         = tmpDir + "/compaction"
+	restoreClusterToken             = "etcd-cluster"
+)
+
+// Compactor holds the necessary details for compacting ETCD
+type Compactor struct {
+	logger *logrus.Entry
+	store  snapstore.SnapStore
+}
+
+// NewCompactor creates compactor
+func NewCompactor(store snapstore.SnapStore, logger *logrus.Entry) *Compactor {
+	return &Compactor{
+		logger: logger,
+		store:  store,
+	}
+}
+
+// Compact is mainly responsible for applying snapshots (full + delta), compacting, drefragmenting, taking the snapshot and saving it sequentially.
+func (cp *Compactor) Compact(ro *brtypes.RestoreOptions, needDefragmentation bool) (*brtypes.CompactionResult, error) {
+	cp.logger.Info("Start compacting")
+
+	// If no basesnapshot is found, abort compaction as there would be nothing to compact
+	if ro.BaseSnapshot == nil {
+		cp.logger.Error("No base snapshot found. Nothing is available for compaction")
+		return nil, fmt.Errorf("No base snapshot found. Nothing is available for compaction")
+	}
+
+	suffix := ro.BaseSnapshot.CompressionSuffix
+
+	// TODO: Remove this provision of saving the compacted snapshot in current directory when we change directory structure
+	cwd, err := os.Getwd()
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create file to save compacted snapshot %v", err)
+	}
+
+	cmpctdFileName := filepath.Join(cwd, filepath.Base(fmt.Sprintf("%s-%d%s", "compacted", time.Now().UTC().Unix(), suffix)))
+	cmpctdFile, err := os.Create(cmpctdFileName)
```

Comment:
[here](https://github.com/gardener/etcd-backup-restore/pull/301/files#diff-adc81f6ffba52bc1871b0dac7c88237aa1cdea40e7691e35e834b40129ab723fR62) you are taking suffix of BaseSnapshot and now you have a filePointer `cmpctdFile` for fileName: `cmpctdFileName` and this fileName have suffix of BaseSnapshot but suffix can be change when deltaSnapList is not empty but [here](https://github.com/gardener/etcd-backup-restore/pull/301/files#diff-adc81f6ffba52bc1871b0dac7c88237aa1cdea40e7691e35e834b40129ab723fR208) you are using the same filePointer `cmpctdFile` for fileName: `cmpctdFileName` to copy the data.

Reply:
Yes. My bad. Fixed it. Thanks, for the catch.